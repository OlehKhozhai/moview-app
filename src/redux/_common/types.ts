import { tabs } from 'config';

export interface ICommonReducer {
  activeTab: typeof tabs[number];
}

export const SET_ACTIVE_TAB = 'SET_ACTIVE_TAB';
export interface ISetCurrentTabAction {
  type: typeof SET_ACTIVE_TAB;
  payload: typeof tabs[number];
}

export type CommonActionTypes = ISetCurrentTabAction;
