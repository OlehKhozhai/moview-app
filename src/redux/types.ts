export type Movie = {
  id: number;
  title: string;
  tagline: string;
  vote_average: number;
  vote_count: number;
  release_date: string;
  poster_path: string;
  overview: string;
  budget: number;
  revenue: number;
  genres: Array<string>;
  runtime: number;
};

export interface RootReducer {
  movies: Movie[];
  isMoviesLoading: boolean;
  movieDetails: Movie | null;
  error: null | string | Array<string>;
}

// GET_MOVIES
export const GET_MOVIES = 'GET_MOVIES';
export const GET_MOVIES_SUCCESS = 'GET_MOVIES_SUCCESS';
export const GET_MOVIES_FAIL = 'GET_MOVIES_FAIL';
export interface GetMoviesAction {
  type: typeof GET_MOVIES;
  payload: {
    params: string;
  };
}
export interface GetMoviesSuccessAction {
  type: typeof GET_MOVIES_SUCCESS;
  payload: Movie[];
}
export interface GetMoviesFailAction {
  type: typeof GET_MOVIES_FAIL;
  payload: string;
}

// GET_MOVIE_DETAILS
export const GET_MOVIE_DETAILS = 'GET_MOVIE_DETAILS';
export const GET_MOVIE_DETAILS_SUCCESS = 'GET_MOVIE_DETAILS_SUCCESS';
export const GET_MOVIE_DETAILS_FAIL = 'GET_MOVIE_DETAILS_FAIL';
export interface GetMovieDetailsAction {
  type: typeof GET_MOVIE_DETAILS;
  payload: number;
}
export interface GetMovieDetailsSuccessAction {
  type: typeof GET_MOVIE_DETAILS_SUCCESS;
  payload: Movie;
}
export interface GetMovieDetailsFailAction {
  type: typeof GET_MOVIE_DETAILS_FAIL;
  payload: string;
}

// CREATE_MOVIE
export const CREATE_MOVIE = 'CREATE_MOVIE';
export const CREATE_MOVIE_SUCCESS = 'CREATE_MOVIE_SUCCESS';
export const CREATE_MOVIE_FAIL = 'CREATE_MOVIE_FAIL';
export interface CreateMovieAction {
  type: typeof CREATE_MOVIE;
  payload: Partial<Movie>;
}
export interface CreateMovieSuccessAction {
  type: typeof CREATE_MOVIE_SUCCESS;
}
export interface CreateMovieFailAction {
  type: typeof CREATE_MOVIE_FAIL;
  payload: string;
}

// EDIT_MOVIE
export const EDIT_MOVIE = 'EDIT_MOVIE';
export const EDIT_MOVIE_SUCCESS = 'EDIT_MOVIE_SUCCESS';
export const EDIT_MOVIE_FAIL = 'EDIT_MOVIE_FAIL';
export interface EditMovieAction {
  type: typeof EDIT_MOVIE;
  payload: Partial<Movie>;
}
export interface EditMovieSuccessAction {
  type: typeof EDIT_MOVIE_SUCCESS;
  payload: Movie;
}
export interface EditMovieFailAction {
  type: typeof EDIT_MOVIE_FAIL;
  payload: string;
}

// REMOVE_MOVIE
export const REMOVE_MOVIE = 'REMOVE_MOVIE';
export const REMOVE_MOVIE_SUCCESS = 'REMOVE_MOVIE_SUCCESS';
export const REMOVE_MOVIE_FAIL = 'REMOVE_MOVIE_FAIL';
export interface RemoveMovieAction {
  type: typeof REMOVE_MOVIE;
  payload: number;
}
export interface RemoveMovieSuccessAction {
  type: typeof REMOVE_MOVIE_SUCCESS;
}
export interface RemoveMovieFailAction {
  type: typeof REMOVE_MOVIE_FAIL;
  payload: string;
}

export type ActionTypes =
  | GetMoviesAction
  | GetMoviesSuccessAction
  | GetMoviesFailAction
  | GetMovieDetailsAction
  | GetMovieDetailsSuccessAction
  | GetMovieDetailsFailAction
  | CreateMovieAction
  | CreateMovieSuccessAction
  | CreateMovieFailAction
  | EditMovieAction
  | EditMovieSuccessAction
  | EditMovieFailAction
  | RemoveMovieAction
  | RemoveMovieSuccessAction
  | RemoveMovieFailAction;
