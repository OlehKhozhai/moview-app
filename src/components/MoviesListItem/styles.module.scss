@import 'assets/styles/mixins/colors';

.root {
  max-width: 330px;
  width: 100%;
  margin-bottom: 50px;
  color: $primary-dark-color;
  position: relative;
  cursor: pointer;
}

.image {
  width: 100%;
}

.titleAndImageWrapper {
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin: 20px 0 15px;
}

.title {
  font-size: 18px;
}

.year {
  padding: 5px 20px;
  border: 1px solid $tertiary-border;
  border-radius: 3px;
  font-size: 14px;
}

.genre {
  font-size: 14px;
}

.dropdown {
  position: absolute;
  top: 10px;
  right: 0;
}
