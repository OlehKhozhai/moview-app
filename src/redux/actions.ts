import {
  GetMoviesSuccessAction,
  GET_MOVIES,
  GET_MOVIES_FAIL,
  GET_MOVIES_SUCCESS,
  GetMoviesAction,
  GetMovieDetailsAction,
  GET_MOVIE_DETAILS,
  GET_MOVIE_DETAILS_SUCCESS,
  GET_MOVIE_DETAILS_FAIL,
  GetMovieDetailsSuccessAction,
  GetMovieDetailsFailAction,
  GetMoviesFailAction,
  CreateMovieAction,
  CREATE_MOVIE_SUCCESS,
  CREATE_MOVIE_FAIL,
  CreateMovieSuccessAction,
  CreateMovieFailAction,
  CREATE_MOVIE,
  RemoveMovieAction,
  EditMovieAction,
  EditMovieFailAction,
  EditMovieSuccessAction,
  EDIT_MOVIE_SUCCESS,
  EDIT_MOVIE_FAIL,
  EDIT_MOVIE,
  RemoveMovieSuccessAction,
  REMOVE_MOVIE_FAIL,
  RemoveMovieFailAction,
  REMOVE_MOVIE,
  REMOVE_MOVIE_SUCCESS,
} from 'redux/types';
import helpers from 'helpers';

export type Dispatch = (data: unknown) => void;

export const getMoviesAction =
  ({ params }: GetMoviesAction['payload']) =>
  async (dispatch: Dispatch) => {
    dispatch({ type: GET_MOVIES });

    return helpers
      .doFetch({ params: `?${params}&sortOrder=desc` })
      .then(({ data }) => {
        dispatch({
          type: GET_MOVIES_SUCCESS,
          payload: data,
        } as GetMoviesSuccessAction);
      })
      .catch((error) => {
        dispatch({
          type: GET_MOVIES_FAIL,
          payload: error,
        } as GetMoviesFailAction);
      });
  };

export const getMovieDetailsAction =
  (id: GetMovieDetailsAction['payload']) => (dispatch: Dispatch) => {
    dispatch({
      type: GET_MOVIE_DETAILS,
      payload: id,
    } as GetMovieDetailsAction);

    return helpers
      .doFetch({ params: `/${id}` })
      .then((movie) => {
        dispatch({
          type: GET_MOVIE_DETAILS_SUCCESS,
          payload: movie,
        } as GetMovieDetailsSuccessAction);
      })
      .catch((error) => {
        dispatch({
          type: GET_MOVIE_DETAILS_FAIL,
          payload: error,
        } as GetMovieDetailsFailAction);
      });
  };

export const createMovieAction =
  (movie: CreateMovieAction['payload']) => async (dispatch: Dispatch) => {
    dispatch({
      type: CREATE_MOVIE,
    });

    return helpers
      .doFetch({ method: 'POST', data: movie })
      .then((createdMovie) => {
        if (createdMovie.messages) {
          dispatch({
            type: CREATE_MOVIE_FAIL,
            payload: createdMovie.messages,
          } as CreateMovieFailAction);

          return createdMovie.messages;
        } else {
          dispatch({
            type: CREATE_MOVIE_SUCCESS,
            payload: createdMovie,
          } as CreateMovieSuccessAction);
        }
      })
      .catch((error) => {
        dispatch({
          type: CREATE_MOVIE_FAIL,
          payload: error,
        } as CreateMovieFailAction);

        return error;
      });
  };

export const editMovieAction =
  (movie: EditMovieAction['payload']) => async (dispatch: Dispatch) => {
    dispatch({
      type: EDIT_MOVIE,
    });

    return helpers
      .doFetch({ method: 'PUT', data: movie })
      .then((updatedMovie) => {
        if (updatedMovie.messages) {
          dispatch({
            type: EDIT_MOVIE_FAIL,
            payload: updatedMovie.messages,
          } as EditMovieFailAction);

          return updatedMovie.messages;
        } else {
          dispatch({
            type: EDIT_MOVIE_SUCCESS,
            payload: updatedMovie,
          } as EditMovieSuccessAction);
        }
      })
      .catch((error) => {
        dispatch({
          type: EDIT_MOVIE_FAIL,
          payload: error,
        } as EditMovieFailAction);
      });
  };

export const removeMovieAction = (id: RemoveMovieAction['payload']) => (dispatch: Dispatch) => {
  dispatch({
    type: REMOVE_MOVIE,
  });

  return helpers
    .doFetch({ method: 'DELETE', params: `/${id}` })
    .then(() => {
      dispatch({
        type: REMOVE_MOVIE_SUCCESS,
      } as RemoveMovieSuccessAction);
    })
    .catch((error) => {
      dispatch({
        type: REMOVE_MOVIE_FAIL,
        payload: error,
      } as RemoveMovieFailAction);
    });
};
