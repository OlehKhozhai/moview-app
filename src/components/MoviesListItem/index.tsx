import React from 'react';
import cn from 'classnames';
import { useHistory } from 'react-router-dom';
import { useDispatch } from 'react-redux';

import useDropdown from 'hooks/useDropdown';
import useOpenAndClose from 'hooks/useOpenAndClose';
import Dropdown from 'components/_common/Dropdown';
import DeleteMovie from 'components/_common/DeleteMovie';
import Modal from 'components/_common/Modal';
import CreateAndEditMovie from 'components/_common/CreateAndEditMovie';
import { MOVIE_OPTIONS } from 'config';
import { Movie } from 'redux/types';
import { removeMovieAction } from 'redux/actions';
import styles from './styles.module.scss';

type MoviesListItemProps = Movie & {
  className?: string;
};

const MoviesListItem: React.FC<MoviesListItemProps> = ({
  className,
  id,
  poster_path,
  title,
  genres,
  release_date,
  runtime,
  overview,
}) => {
  const dispatch = useDispatch();
  const history = useHistory();
  const { value, isDropdownVisible, onOptionClick, onToggleDropdownVisibility, onCloseDropdown } =
    useDropdown();
  const {
    isOpen: isOpenDeleteMovieModal,
    onClose: onCloseDeleteMovieModal,
    onOpen: onOpenDeleteMovieModal,
  } = useOpenAndClose({ onCloseCallback: () => onOptionClick('') });
  const {
    isOpen: isOpenEditMovieModal,
    onClose: onCloseEditMovieModal,
    onOpen: onOpenEditMovieModal,
  } = useOpenAndClose({ onCloseCallback: () => onOptionClick('') });

  const handleClickOnMovie = () => {
    history.push(`/details/${id}`);
    window.scrollTo({ behavior: 'smooth', top: 0 });
  };

  const handleOnConfirmMovieRemove = React.useCallback(() => {
    onCloseDeleteMovieModal();
    dispatch(removeMovieAction(id));
  }, [id, dispatch, onCloseDeleteMovieModal]);

  React.useEffect(() => {
    if (value === 'edit') {
      onOpenEditMovieModal();
    }
    if (value === 'delete') {
      onOpenDeleteMovieModal();
    }
  }, [value, onOpenEditMovieModal, onOpenDeleteMovieModal]);

  return (
    <>
      <li className={cn(styles.root, className)} onClick={handleClickOnMovie}>
        <Dropdown
          options={MOVIE_OPTIONS}
          isOpen={isDropdownVisible}
          onToggle={onToggleDropdownVisibility}
          onClose={onCloseDropdown}
          onOptionClick={onOptionClick}
          className={styles.dropdown}
        />

        <img src={poster_path} alt={title} className={styles.image} />

        <div className={styles.titleAndImageWrapper}>
          <h4 className={styles.title}>{title}</h4>
          <span className={styles.year}>{release_date}</span>
        </div>

        <h5 className={styles.genre}>
          {genres.map((genre) => (
            <span key={genre}> {genre} |</span>
          ))}
        </h5>
      </li>

      {isOpenDeleteMovieModal && (
        <Modal isOpen={isOpenDeleteMovieModal} onClose={onCloseDeleteMovieModal}>
          <DeleteMovie onConfirm={handleOnConfirmMovieRemove} />
        </Modal>
      )}

      {isOpenEditMovieModal && (
        <Modal isOpen={isOpenEditMovieModal} onClose={onCloseEditMovieModal}>
          <CreateAndEditMovie
            action="edit"
            movieId={id}
            initialValues={{
              title,
              genres,
              overview,
              poster_path,
              release_date,
              runtime: runtime || 0,
            }}
            onSubmit={onCloseEditMovieModal}
          />
        </Modal>
      )}
    </>
  );
};

export default React.memo(MoviesListItem);
