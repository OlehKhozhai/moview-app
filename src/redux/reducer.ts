import {
  GET_MOVIES_FAIL,
  GET_MOVIES_SUCCESS,
  GET_MOVIE_DETAILS,
  GET_MOVIE_DETAILS_FAIL,
  GET_MOVIE_DETAILS_SUCCESS,
  CREATE_MOVIE,
  CREATE_MOVIE_FAIL,
  CREATE_MOVIE_SUCCESS,
  EDIT_MOVIE,
  EDIT_MOVIE_FAIL,
  EDIT_MOVIE_SUCCESS,
  REMOVE_MOVIE,
  REMOVE_MOVIE_SUCCESS,
  REMOVE_MOVIE_FAIL,
} from './types';
import { ActionTypes, GET_MOVIES, RootReducer } from 'redux/types';

export const initialState: RootReducer = {
  movies: [],
  isMoviesLoading: false,
  movieDetails: null,
  error: null,
};

export default (state = initialState, action: ActionTypes): RootReducer => {
  switch (action.type) {
    case GET_MOVIES:
      return {
        ...state,
        isMoviesLoading: true,
        error: null,
      };
    case GET_MOVIES_SUCCESS:
      return {
        ...state,
        movies: action.payload,
        isMoviesLoading: false,
      };
    case GET_MOVIES_FAIL:
      return {
        ...state,
        movies: [],
        isMoviesLoading: false,
        error: action.payload,
      };

    case GET_MOVIE_DETAILS:
      return { ...state, error: null };
    case GET_MOVIE_DETAILS_SUCCESS:
      return {
        ...state,
        movieDetails: action.payload,
      };
    case GET_MOVIE_DETAILS_FAIL:
      return {
        ...state,
        movieDetails: null,
        error: action.payload,
      };

    case CREATE_MOVIE:
    case EDIT_MOVIE:
    case REMOVE_MOVIE:
    case REMOVE_MOVIE_SUCCESS:
    case CREATE_MOVIE_SUCCESS:
      return { ...state, error: null };
    case EDIT_MOVIE_SUCCESS:
      return {
        ...state,
        movies: state.movies.map((movie) =>
          movie.id === action.payload.id ? action.payload : movie
        ),
        movieDetails:
          state.movieDetails?.id === action.payload.id ? action.payload : state.movieDetails,
      };
    case CREATE_MOVIE_FAIL:
    case EDIT_MOVIE_FAIL:
    case REMOVE_MOVIE_FAIL:
      return {
        ...state,
        error: action.payload,
      };

    default:
      return state;
  }
};
