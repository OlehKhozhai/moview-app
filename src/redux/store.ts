import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';

import rootReducer from 'redux/reducer';

const middleware = [thunk];
const enhancer = composeWithDevTools(applyMiddleware(...middleware));

export type AppStore = ReturnType<typeof rootReducer>;

export const store = createStore(rootReducer, enhancer);
