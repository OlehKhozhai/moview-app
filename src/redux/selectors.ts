import { AppStore } from 'redux/store';

export const moviesSelector = (store: AppStore) => ({
  movies: store.movies,
  isMoviesLoading: store.isMoviesLoading,
});

export const movieDetailsSelector = (store: AppStore) => store.movieDetails;

export const errorSelector = (store: AppStore) => store.error;
