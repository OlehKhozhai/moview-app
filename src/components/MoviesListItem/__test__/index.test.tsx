import React from 'react';
import { create } from 'react-test-renderer';
import rrd from 'react-router-dom';

import { Movie } from 'redux/types';
import MoviesListItem from '../index';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));

const movie: Movie = {
  id: 1,
  budget: 100,
  genres: ['Documentary'],
  overview: 'Overview',
  poster_path: 'https://via.placeholder.com/330x450',
  title: 'Movie 1',
  release_date: '1999',
  revenue: 102,
  runtime: 12312,
  tagline: 'tagline',
  vote_average: 4,
  vote_count: 100,
};

const { MemoryRouter, Router } = rrd;

describe('MoviesListItem component', () => {
  test('should render MoviesListItem component by default', () => {
    const component = create(
      <MemoryRouter initialEntries={['/']}>
        <MoviesListItem {...movie} />
      </MemoryRouter>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should redirect to movie details page', () => {
    const useHistoryPushStub = jest.fn();
    const scrollToStub = jest.fn();
    global.scrollTo = scrollToStub;

    const component = create(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      <Router history={{ push: useHistoryPushStub, location: {}, listen: jest.fn() } as any}>
        <MoviesListItem {...movie} />
      </Router>
    );

    component.root.findAllByType('li')[0].props.onClick();

    expect(useHistoryPushStub).toHaveBeenCalledWith('/movies/1');
    expect(scrollToStub).toHaveBeenCalled();
  });
});
