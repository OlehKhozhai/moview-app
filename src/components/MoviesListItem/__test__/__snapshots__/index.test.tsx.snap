// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesListItem component should render MoviesListItem component by default 1`] = `
<li
  className="root"
  onClick={[Function]}
>
  <div
    className="root dropdown"
  >
    <span
      className="threeDots"
      onClick={[Function]}
    >
      ...
    </span>
  </div>
  <img
    alt="Movie 1"
    className="image"
    src="https://via.placeholder.com/330x450"
  />
  <div
    className="titleAndImageWrapper"
  >
    <h4
      className="title"
    >
      Movie 1
    </h4>
    <span
      className="year"
    >
      1999
    </span>
  </div>
  <h5
    className="genre"
  >
    <span>
       
      Documentary
       |
    </span>
  </h5>
</li>
`;
